metals/src/main/scala/scala/meta/internal/decorations/SyntheticsDecorationProvider.scala

```diff
@@ -298,26 +318,147 @@ final class SyntheticsDecorationProvider(
   private def publishSyntheticDecorations(
       path: AbsolutePath,
       textDocument: TextDocument
-  ): Unit =
+  ): Unit = {
     if (clientConfig.isInlineDecorationProvider()) {
-      val edit = buffer.tokenEditDistance(path, textDocument.text)
+
+      lazy val edit = buffer.tokenEditDistance(path, textDocument.text)
+
+      val decorationPrinter = new SemanticdbTreePrinter(
+        isHover = false,
+        toDecorationString(textDocument),
+        PrinterSymtab.fromTextDocument(textDocument)
+      )
+
       val decorations = for {
         synthetic <- textDocument.synthetics
-        (decoration, range) <- printSyntheticInfo(
-          textDocument,
-          synthetic,
-          toDecorationString(textDocument),
-          userConfig(),
-          isHover = false
-        )
-        lspRange <- edit.toRevisedStrict(range).toIterable
+        (decoration, range) <- decorationPrinter
+          .printSyntheticInfo(
+            textDocument,
+            synthetic,
+            userConfig()
+          )
+          .toIterable
+        currentRange <- edit.toRevisedStrict(range).toIterable
+        lspRange = currentRange.toLSP
       } yield decorationOptions(lspRange, decoration)
 
+      val typDecorations =
+        if (userConfig().showInferredType)
+          typeDecorations(path, textDocument, decorationPrinter)
+        else Nil
       val params =
         new PublishDecorationsParams(
           path.toURI.toString(),
-          decorations.toArray
+          (decorations ++ typDecorations).toArray
         )
+
       client.metalsPublishDecorations(params)
     }
+  }
+
+  private def typeDecorations(
+      path: AbsolutePath,
+      textDocument: TextDocument,
+      decorationPrinter: SemanticdbTreePrinter
+  ) = {
+
+    val methodPositions = mutable.Map.empty[s.Range, s.Range]
+
+    def explorePatterns(pats: List[m.Pat]): List[s.Range] = {
+      pats.flatMap {
+        case m.Pat.Var(nm @ m.Term.Name(_)) =>
+          List(nm.pos.toSemanticdb)
+        case m.Pat.Extract((_, pats)) =>
+          explorePatterns(pats)
+        case m.Pat.ExtractInfix(lhs, _, pats) =>
+          explorePatterns(lhs :: pats)
+        case m.Pat.Tuple(tuplePats) =>
+          explorePatterns(tuplePats)
+        case m.Pat.Bind(lhs, rhs) =>
+          List(lhs.pos.toSemanticdb) ++ explorePatterns(List(rhs))
+        case _ => Nil
+      }
+    }
+
+    def visit(tree: m.Tree): List[s.Range] = {
+      tree match {
+        case vl: m.Defn.Val =>
+          val values =
+            if (vl.decltpe.isEmpty) explorePatterns(vl.pats) else Nil
+          values ++ visit(vl.rhs)
```

Comment:
What does `rhs` stand for? I know it's a `Term`, but am a bit unfamiliar with it?

Reply:
That's `right hand side` - it's heavily used in the parser, maybe too heavily :sweat_smile:

Reply:
ah of course. I figured it was something super obvious I wasn't thinking of 👍🏼